File: 02-jetson/02-PC-web_viewer.py
import cv2
import zmq
import base64
import numpy as np
import time
from flask import Flask, render_template_string, Response
from datetime import datetime

# --- CONFIGURATION ---
JETSON_IP = "192.168.37.22"
ZMQ_PORT = 5555
THRESHOLD = 0.70

app = Flask(__name__)

PAGE = """
<html>
<head><title>JetsonPilot Dashboard</title></head>
<body style="background: #222; color: #eee; text-align: center; font-family: monospace;">
    <h1>JetsonPilot Real-Time View</h1>
    <div style="margin-bottom: 10px; color: #aaa;">Data source: Jetson Nano (ResNet18)</div>
    <img src="/video_feed" style="border: 2px solid #444; width: 672px; image-rendering: pixelated;"/>
</body>
</html>
"""

class VideoProcessor:
    def __init__(self):
        self.context = zmq.Context()
        self.socket = self.context.socket(zmq.SUB)
        # Low Conflate means: "Give me only the newest message, drop old ones"
        # Great to avoid the "250 FPS buffer read" issue you had
        self.socket.setsockopt(zmq.CONFLATE, 1)

        print(f"[Connect] Connecting to Jetson at {JETSON_IP}:{ZMQ_PORT}...")
        self.socket.connect(f"tcp://{JETSON_IP}:{ZMQ_PORT}")
        self.socket.setsockopt_string(zmq.SUBSCRIBE, "")

        # Recording setup
        timestamp = datetime.now().strftime("%Y%m%d_%H%M%S")
        self.filename = f"jetson_recording_{timestamp}.avi"
        self.writer = None
        print(f"[Record] Saving to {self.filename}")

    def process_stream(self):
        while True:
            try:
                # 1. Receive Data
                # Since we use CONFLATE, this always gives the absolute latest frame
                data = self.socket.recv_json()

                prob = data['prob_target']
                fps = data['jetson_fps'] # Retrieve the REAL FPS from Jetson
                jpg_original = base64.b64decode(data['image_b64'])

                # 2. Decode
                np_arr = np.frombuffer(jpg_original, dtype=np.uint8)
                frame = cv2.imdecode(np_arr, cv2.IMREAD_COLOR)

                if frame is None: continue

                # 3. Draw Overlay
                if prob > THRESHOLD:
                    label = f"CIBLE:{prob:.0%}"
                    color = (0, 255, 0)
                    bg_color = (0, 50, 0)
                else:
                    label = f"NOCIBLE:{prob:.0%}"
                    color = (0, 0, 255)
                    bg_color = (0, 0, 50)

                # Top Bar
                cv2.rectangle(frame, (0, 0), (frame.shape[1], 20), bg_color, -1)

                # Info Text
                font = cv2.FONT_HERSHEY_SIMPLEX
                # Target Info
                cv2.putText(frame, label, (5, 15), font, 0.4, color, 1)

                # FPS Info (From Jetson)
                fps_text = f"INF:{fps:.1f} FPS"
                (w, _), _ = cv2.getTextSize(fps_text, font, 0.4, 1)
                cv2.putText(frame, fps_text, (frame.shape[1] - w - 5, 15), font, 0.4, (200, 200, 200), 1)

                # 4. Save to Disk
                if self.writer is None:
                    h, w_img = frame.shape[:2]
                    self.writer = cv2.VideoWriter(self.filename, cv2.VideoWriter_fourcc(*'MJPG'), 20.0, (w_img, h))
                self.writer.write(frame)

                # 5. Browser Stream
                _, buffer = cv2.imencode('.jpg', frame)
                yield (b'--frame\r\n'
                       b'Content-Type: image/jpeg\r\n\r\n' + buffer.tobytes() + b'\r\n')

            except Exception as e:
                # print(f"Error: {e}")
                pass

processor = VideoProcessor()

@app.route('/')
def index():
    return render_template_string(PAGE)

@app.route('/video_feed')
def video_feed():
    return Response(processor.process_stream(), mimetype='multipart/x-mixed-replace; boundary=frame')

if __name__ == '__main__':
    app.run(host='0.0.0.0', port=5000, debug=False, threaded=True)

File: 02-jetson/02-PC-web_viewer.go
package main

import (
	"encoding/base64"
	"encoding/json"
	"fmt"
	"image"
	"image/color"
	"io/ioutil"
	"log"
	"net/http"
	"os"
	"path/filepath"
	"sync"
	"time"

	zmq "github.com/pebbe/zmq4"
	"gocv.io/x/gocv"
)

// --- CONFIGURATION ---
const (
	JetsonIP   = "192.168.37.22"
	ZmqPort    = "5555"
	DatasetDir = "dataset_capture"
	Threshold  = 0.70
)

// --- GLOBAL STATE ---
var (
	// Stores the latest RAW jpeg bytes from Jetson (Clean image for photos)
	latestRawImage []byte
	imgMutex       sync.RWMutex

	// Channel to send processed images (with overlay) to the web stream
	streamChannel = make(chan []byte, 1)
)

// Data structure matches the JSON sent by Jetson
type VisionData struct {
	ProbTarget float64 `json:"prob_target"`
	ImageB64   string  `json:"image_b64"`
	FPS        float64 `json:"jetson_fps"`
}

// HTML Interface
const PAGE = `
<html>
<head>
    <title>Jetson Pilot</title>
    <style>
        body { background: #222; color: #eee; text-align: center; font-family: monospace; overflow: hidden; }
        #container { position: relative; display: inline-block; border: 2px solid #444; margin-top: 20px; }
        img { width: 672px; image-rendering: pixelated; display: block; }
        #flash {
            position: absolute; top: 0; left: 0; width: 100%; height: 100%;
            background: white; opacity: 0; pointer-events: none; transition: opacity 0.1s;
        }
        #log { margin-top: 15px; color: #4CAF50; font-weight: bold; height: 20px; }
    </style>
</head>
<body>
    <h1>Jetson Pilot</h1>
    <div style="color: #aaa;">[SPACE] or CLICK to Save Raw Photo</div>

    <div id="container" onclick="capture()">
        <div id="flash"></div>
        <img src="/video_feed" />
    </div>
    <div id="log">Recording video... Ready.</div>

    <script>
        function capture() {
            const flash = document.getElementById('flash');
            flash.style.opacity = 0.8;
            setTimeout(() => { flash.style.opacity = 0; }, 100);

            fetch('/capture_trigger')
                .then(response => response.text())
                .then(msg => {
                    const log = document.getElementById('log');
                    log.innerText = msg;
                    log.style.opacity = 1;
                    setTimeout(() => { log.style.opacity = 0.5; }, 2000);
                })
                .catch(err => console.error(err));
        }

        document.addEventListener('keydown', function(event) {
            if (event.code === 'Space') {
                event.preventDefault();
                capture();
            }
        });
    </script>
</body>
</html>
`

func main() {
	// 1. Prepare Directory
	if _, err := os.Stat(DatasetDir); os.IsNotExist(err) {
		os.MkdirAll(DatasetDir, 0755)
	}

	// 2. Start ZMQ Logic
	go zmqLoop()

	// 3. Start Web Server
	http.HandleFunc("/", func(w http.ResponseWriter, r *http.Request) { w.Write([]byte(PAGE)) })
	http.HandleFunc("/video_feed", streamHandler)
	http.HandleFunc("/capture_trigger", captureHandler)

	fmt.Println("[System] Collector running on :5000")
	fmt.Printf("[Info]  Photos -> '%s/'\n", DatasetDir)
	log.Fatal(http.ListenAndServe(":5000", nil))
}

func zmqLoop() {
	// Setup ZMQ
	zCtx, _ := zmq.NewContext()
	socket, _ := zCtx.NewSocket(zmq.SUB)
	defer socket.Close()

	addr := fmt.Sprintf("tcp://%s:%s", JetsonIP, ZmqPort)
	fmt.Printf("[Connect] Connecting to %s...\n", addr)
	socket.Connect(addr)
	socket.SetSubscribe("")
	socket.SetConflate(true)

	// Setup Video Writer (Using OpenCV - Real AVI)
	timestamp := time.Now().Format("20060102_150405")
	filename := fmt.Sprintf("recording_%s.avi", timestamp)

	// MJPG is a safe bet for AVI container without complex codec installs
	writer, err := gocv.VideoWriterFile(filename, "MJPG", 20, 224, 224, true)
	if err != nil {
		fmt.Printf("[Error] Could not open video writer: %v\n", err)
	} else {
		fmt.Printf("[Record] Saving video to %s\n", filename)
		defer writer.Close()
	}

	for {
		msg, err := socket.Recv(0)
		if err != nil {
			continue
		}

		var data VisionData
		if err := json.Unmarshal([]byte(msg), &data); err != nil {
			continue
		}

		// Decode Base64
		rawBytes, err := base64.StdEncoding.DecodeString(data.ImageB64)
		if err != nil {
			continue
		}

		// 1. Store RAW bytes for Photo Capture (Thread Safe)
		imgMutex.Lock()
		latestRawImage = rawBytes
		imgMutex.Unlock()

		// 2. Decode to OpenCV Mat for Processing
		img, err := gocv.IMDecode(rawBytes, gocv.IMReadColor)
		if err != nil {
			continue
		}
		if img.Empty() {
			continue
		}

		// --- DRAW OVERLAY (Using OpenCV - Clean Text) ---
		var label string
		var rectCol color.RGBA
		// var textCol color.RGBA

		if data.ProbTarget > Threshold {
			label = fmt.Sprintf("CIBLE:%.0f%%", data.ProbTarget*100)
			// textCol = color.RGBA{0, 255, 0, 0} // Green Text
			rectCol = color.RGBA{0, 50, 0, 0} // Dark Green BG
		} else {
			label = fmt.Sprintf("NOCIBLE:%.0f%%", data.ProbTarget*100)
			// textCol = color.RGBA{0, 0, 255, 0} // Red Text
			rectCol = color.RGBA{0, 0, 50, 0} // Dark Red BG
		}

		// Draw Top Bar Background
		gocv.Rectangle(&img, image.Rect(0, 0, 224, 20), rectCol, -1)

		// Draw Labels
		// FONT_HERSHEY_SIMPLEX ensures the ":" renders correctly
		gocv.PutText(&img, label, image.Pt(5, 15), gocv.FontHersheySimplex, 0.4, color.RGBA{200, 200, 200, 0}, 1)

		fpsText := fmt.Sprintf("FPS:%.1f", data.FPS)
		gocv.PutText(&img, fpsText, image.Pt(150, 15), gocv.FontHersheySimplex, 0.4, color.RGBA{200, 200, 200, 0}, 1)

		// 3. Write Frame to Video File
		if writer.IsOpened() {
			writer.Write(img)
		}

		// 4. Encode to JPEG for Web Stream
		buf, _ := gocv.IMEncode(".jpg", img)
		select {
		case streamChannel <- buf.GetBytes():
		default:
		}

		// Important: Release OpenCV memory manually
		img.Close()
	}
}

// --- HANDLERS ---

func captureHandler(w http.ResponseWriter, r *http.Request) {
	imgMutex.RLock()
	data := latestRawImage // Get the CLEAN bytes
	imgMutex.RUnlock()

	if data == nil {
		w.WriteHeader(http.StatusServiceUnavailable)
		return
	}

	// Generate filename
	files, _ := ioutil.ReadDir(DatasetDir)
	id := len(files)
	filename := fmt.Sprintf("%d.jpg", id)
	path := filepath.Join(DatasetDir, filename)

	// Save RAW data
	if err := ioutil.WriteFile(path, data, 0644); err != nil {
		w.Write([]byte("Error Saving"))
		return
	}

	msg := fmt.Sprintf("Saved: %s", filename)
	fmt.Printf("[Photo] %s\n", msg)
	w.Write([]byte(msg))
}

func streamHandler(w http.ResponseWriter, r *http.Request) {
	w.Header().Set("Content-Type", "multipart/x-mixed-replace; boundary=frame")
	for jpgData := range streamChannel {
		fmt.Fprintf(w, "--frame\r\nContent-Type: image/jpeg\r\n\r\n%s\r\n", jpgData)
	}
}
